### DISCUSSION TITLE
Dynamic columns with Scipted Data Source

### DISCUSSION THREAD

**Discussion by BobanNikolov** [eclipse-birt/birt#1355] [Q&A]
Hi, the issue is following.
I am using the following BIRT version : org.eclipse.birt.runtime_4.8.0-20180626
I have a requirement where I need to make dynamic columns. There are required fields which are written in the application and for each one of them there needs to be a separate column with the parameter name as a header and the parameter value as a info row. I have knowledge only with Scripted Data Source, but the struggle has been real :) 
Please provide me with a solution
P.S. Every kind of data migration will work except JDBC connection.

**Comment by hvbtup** [comment]
Your description isn't really clear.

My interpretation of what you want is:

There are N data records.
The report should output a table with a header row and N detail rows. 
Until here this would be easy: Use a table item.
However, the number of columns, and which column should show which attribute of a record, is dynamic.
In other word, the sequence of attribute names and their correspondings titles is dynamic.
E.g. a "persons" report might output the columns ["name", "birthday"] now and ["gender", "name", "weight"] for a different call.

I think there are 2 potential approaches:

1) You could use a crosstab item.
2) You use a table item with the maximum possible number of columns you expect. Then use visibility expressions to hide the columns that you don't need.

A little more info regarding the second approach:

Assume that in your report there is an array variable "columns" which you initialize dynamically e.g. with 

    vars["columns"] = ["gender", "name", "weight"];

In your table item, you would hide all columns (uising a visiblity expression) where the (1-based) column index is >= columns.length. For the other columns (that are not hidden), you need dynamic text items for the titles and for the values.
E.g. for the output in the third column you would use the dynamic text expression

    row[vars["columns"][2]] 

(2 instead of 3 because Javascript arrays are 0-based).

The column titles in the header row can be handled in a similar way (e.g. if you have a variable column_titles = ["Geschlecht", "Name", "Gewicht"], you could use the following expression for the 3rd column:

    vars["column_titles"][2]

to show the titel "Gewicht".
